sweep/sweeper.go

```diff
@@ -1,24 +1,59 @@
 package sweep
 
 import (
+	"errors"
+	"fmt"
 	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
 	"github.com/btcsuite/btcd/txscript"
 	"github.com/btcsuite/btcd/wire"
 	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/chainntnfs"
 	"github.com/lightningnetwork/lnd/lnwallet"
+	"github.com/lightningnetwork/lnd/ticker"
+	"sync"
+	"sync/atomic"
+	"time"
 )
 
-// UtxoSweeper provides the functionality to generate sweep txes. The plan is
-// to extend UtxoSweeper in the future to also manage the actual sweeping
-// process by itself.
+var (
+	// DefaultSweepCheckInterval specifies the time intervals between checks
+	// for inputs to sweep.
+	DefaultSweepCheckInterval = time.Second
+
+	// DefaultBatchWindowSize specifies the number of sweep check intervals
+	// to pass without new inputs being added before the sweep is published.
+	DefaultBatchWindowSize = 30
+
+	// ErrRemoteSpend is returned in case an output that we try to sweep is
+	// confirmed in a tx of the remote party.
+	ErrRemoteSpend = errors.New("remote party swept utxo")
+
+	// sweeperTimeout is the maximum time a user of the sweeper is allowed
+	// to wait before reading the sweep result. This is limited to prevent
+	// blocking of the sweeper.
+	sweeperTimeout = 5 * time.Second
+)
+
+// UtxoSweeper is responsible for sweeping outputs back into the wallet
 type UtxoSweeper struct {
-	cfg *UtxoSweeperConfig
+	cfg  *UtxoSweeperConfig
+	quit chan struct{}
+
+	wg      sync.WaitGroup
+	stopped uint32
+	started uint32
+
+	newInputs chan *sweeperInput
+	spendChan chan *chainntnfs.SpendDetail
```

Comment:
Won't it need to have a spend chan for each output swept?

Reply:
This is a multiple sender, single receiver channel. It is to synchronize all spend notifications to the sweeper main loop.